From 82a17c6653ca9882c71bc51a450f1d38c1bcbafe Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Mon, 26 Aug 2024 14:09:25 +0200
Subject: [PATCH 1098/1695] drivers: stm32_rifsc: restrain access on non secure
 peripherals for OP-TEE

As per previous behavior for the firewall bus probe, do not allow non
secure peripherals to be probed in OP-TEE except for some listed
exceptions. The firewall mechanism exception only applies to the secure
level as bad CID filtering configuration is prohibited by hardware.

To do so, implement a driver specific firewall bus probe that will
only probe secure peripherals and implement firewall exceptions for
which no firewall operations will be done when CFG_INSECURE is set.
This allows, for example, to share a console with the non-secure world
for development purposes.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 59d818d0ffc0c855aefbaa3933af508cad2a611e drivers: stm32_rifsc: restrain access on non secure peripherals for OP-TEE]
Change-Id: I435abe028da41bf6ce2f05b52fbd5c441e22428d
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/400276
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/arch/arm/plat-stm32mp2/main.c       | 18 ++++++
 core/arch/arm/plat-stm32mp2/stm32_util.h |  2 +
 core/drivers/firewall/stm32_rifsc.c      | 79 +++++++++++++++++++++++-
 3 files changed, 98 insertions(+), 1 deletion(-)

diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index 770b2f1bb..4df4d8203 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -469,3 +469,21 @@ void stm32_debug_suspend(unsigned long a0)
 
 	clk_disable(dbgmcu_clk);
 }
+
+bool stm32mp_allow_probe_shared_device(const void *fdt, int node)
+{
+	static int uart_console_node = -1;
+	static bool once;
+
+	if (!once) {
+		get_console_node_from_dt((void *)fdt, &uart_console_node,
+					 NULL, NULL);
+		once = true;
+	}
+
+	/* Allow OP-TEE console to be shared with non-secure world */
+	if (node == uart_console_node)
+		return true;
+
+	return false;
+}
diff --git a/core/arch/arm/plat-stm32mp2/stm32_util.h b/core/arch/arm/plat-stm32mp2/stm32_util.h
index 5c73d95a7..cbacdd526 100644
--- a/core/arch/arm/plat-stm32mp2/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp2/stm32_util.h
@@ -43,4 +43,6 @@ static inline bool stm32_stpmic2_is_present(void)
 
 void stm32_debug_suspend(unsigned long a0);
 
+bool stm32mp_allow_probe_shared_device(const void *fdt, int node);
+
 #endif /*__STM32_UTIL_H__*/
diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index d201ca7ba..1c77d05b9 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -16,6 +16,7 @@
 #include <kernel/pm.h>
 #include <libfdt.h>
 #include <mm/core_memprot.h>
+#include <stm32_util.h>
 #include <tee_api_defines.h>
 #include <trace.h>
 #include <util.h>
@@ -843,8 +844,11 @@ static TEE_Result stm32_rifsc_acquire_access(struct firewall_query *firewall)
 {
 	uintptr_t rifsc_base = rifsc_pdata.base;
 	unsigned int cid_reg_offset = 0;
+	unsigned int periph_offset = 0;
 	unsigned int resource_id = 0;
+	unsigned int reg_id = 0;
 	uint32_t cidcfgr = 0;
+	uint32_t seccfgr = 0;
 
 	assert(rifsc_base);
 
@@ -860,10 +864,17 @@ static TEE_Result stm32_rifsc_acquire_access(struct firewall_query *firewall)
 	if (resource_id >= RIMU_ID_OFFSET)
 		return TEE_SUCCESS;
 
+	reg_id = resource_id / _PERIPH_IDS_PER_REG;
+	periph_offset = resource_id % _PERIPH_IDS_PER_REG;
+
 	cid_reg_offset = _OFFSET_PERX_CIDCFGR * resource_id;
 	cidcfgr = io_read32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
 			    cid_reg_offset);
 
+	seccfgr = io_read32(rifsc_base + _RIFSC_RISC_SECCFGR0 + 0x4 * reg_id);
+	if (!(seccfgr & BIT(periph_offset)))
+		return TEE_ERROR_ACCESS_DENIED;
+
 	/* Only check CID attributes */
 	if (!(cidcfgr & _CIDCFGR_CFEN))
 		return TEE_SUCCESS;
@@ -1061,6 +1072,72 @@ TEE_Result stm32_rifsc_check_tdcid(bool *tdcid_state)
 	return TEE_SUCCESS;
 }
 
+static TEE_Result
+stm32_rifsc_dt_probe_bus(const void *fdt, int node,
+			 struct firewall_controller *ctrl __maybe_unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct firewall_query *fw = NULL;
+	int subnode = 0;
+
+	DMSG("Populating %s firewall bus", ctrl->name);
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		unsigned int i = 0;
+
+		if (fdt_get_status(fdt, subnode) == DT_STATUS_DISABLED)
+			continue;
+
+		if (IS_ENABLED(CFG_INSECURE) &&
+		    stm32mp_allow_probe_shared_device(fdt, subnode)) {
+			DMSG("Skipping firewall attributes check for %s",
+			     fdt_get_name(fdt, subnode, NULL));
+			goto skip_check;
+		}
+
+		DMSG("Acquiring firewall access for %s when probing bus",
+		     fdt_get_name(fdt, subnode, NULL));
+
+		do {
+			/*
+			 * The access-controllers property is mandatory for
+			 * firewall bus devices
+			 */
+			res = firewall_dt_get_by_index(fdt, subnode, i, &fw);
+			if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+				/* Stop when nothing more to parse */
+				break;
+			} else if (res) {
+				EMSG("%s: Error on node %s: %#"PRIx32,
+				     ctrl->name,
+				     fdt_get_name(fdt, subnode, NULL), res);
+				panic();
+			}
+
+			res = firewall_acquire_access(fw);
+			if (res) {
+				EMSG("%s: %s not accessible: %#"PRIx32,
+				     ctrl->name,
+				     fdt_get_name(fdt, subnode, NULL), res);
+				panic();
+			}
+
+			firewall_put(fw);
+			i++;
+		} while (true);
+
+skip_check:
+		res = dt_driver_maybe_add_probe_node(fdt, subnode);
+		if (res) {
+			EMSG("Failed on node %s with %#"PRIx32,
+			     fdt_get_name(fdt, subnode, NULL), res);
+			panic();
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
 static const struct firewall_controller_ops firewall_ops = {
 	.set_conf = stm32_rifsc_set_config,
 	.check_access = stm32_rifsc_check_access,
@@ -1111,7 +1188,7 @@ static TEE_Result stm32_rifsc_probe(const void *fdt, int node,
 	if (res)
 		panic();
 
-	res = firewall_dt_probe_bus(fdt, node, controller);
+	res = stm32_rifsc_dt_probe_bus(fdt, node, controller);
 	if (res)
 		panic();
 
-- 
2.47.2

